### PR TITLE
Buffered connection management for collator-protocol

### PR DESCRIPTION
**Pull request by slumber**
Resolves https://github.com/paritytech/polkadot/issues/5062
Might help with #4617
Needed for #5054 (easily rebased on the feature branch thus targets master)

### CODE REVIEW

File: node/network/collator-protocol/src/collator_side/mod.rs [hunk 1]
```diff
@@ -1023,6 +967,13 @@ pub(crate) async fn run<Context>(
 				FromOrchestra::Signal(Conclude) => return Ok(()),
 			},
 			(relay_parent, peer_id) = state.active_collation_fetches.select_next_some() => {
+				// Schedule a bit reset for this peer.
+				if let Some(authority_ids) = state.peer_ids.get(&peer_id) {
+					state.reset_bit_delays.push(ResetBitDelay::new(
+						relay_parent, authority_ids.clone(), RESET_BIT_DELAY
+					));
+				}
+
```
Comment (slumber) [inline node/network/collator-protocol/src/collator_side/mod.rs:976]: This timeout is probably unnecessary because we already scheduled one on advertisement.
Though we could set this one to be shorter

**Comment by slumber** [comment]
Connected validators are updated on each new advertisement the same way as before,
this way we keep the connection to the current group alive.

For async backing there's one thing to keep in mind: a collation doesn't leave the view immediately once
the leaf gets deactivated. If a validator doesn't fetch a collation within some timeout, we should reset the
corresponding bit ourselves and disconnect if there're no more collations to send

File: node/network/collator-protocol/src/collator_side/validators_buffer.rs [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+/// Unique identifier of a validators group.
+#[derive(Debug, Default, Clone, PartialEq, Eq)]
+struct ValidatorsGroupInfo {
+	len: usize,
+	session_index: SessionIndex,
+	group_index: GroupIndex,
+}
+
+#[derive(Debug)]
+pub struct ValidatorGroupsBuffer<const N: usize> {
```
Comment (rphmeier) [inline node/network/collator-protocol/src/collator_side/validators_buffer.rs:51]: style: we require docs on all public items in any module

File: node/network/collator-protocol/src/collator_side/validators_buffer.rs [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+	buf: VecDeque<ValidatorsGroupInfo>,
+	validators: VecDeque<AuthorityDiscoveryId>,
+	should_be_connected: HashMap<Hash, BitVec>,
+}
+
+impl<const N: usize> ValidatorGroupsBuffer<N> {
+	pub fn new() -> Self {
+		assert!(N > 0);
```
Comment (rphmeier) [inline node/network/collator-protocol/src/collator_side/validators_buffer.rs:59]: panics in production code which are not provably impossible aren't allowed

File: node/network/collator-protocol/src/collator_side/validators_buffer.rs [hunk 4]
```diff
@@ -0,0 +1,265 @@
...
+				bits.set(idx, false);
+			}
+		}
+	}
+
+	pub fn remove_relay_parent(&mut self, relay_parent: &Hash) {
+		self.should_be_connected.remove(relay_parent);
+	}
+
+	fn push(
```
Comment (rphmeier) [inline node/network/collator-protocol/src/collator_side/validators_buffer.rs:123]: it doesn't look like this is ever used. is that not an issue?

File: node/network/collator-protocol/src/collator_side/validators_buffer.rs (hunk 3 above)
Comment (slumber) [reply node/network/collator-protocol/src/collator_side/validators_buffer.rs]: 70bffc7365777089f6f84af524cb0c2938cbeddb

File: node/network/collator-protocol/src/collator_side/validators_buffer.rs (hunk 2 above)
Comment (slumber) [reply node/network/collator-protocol/src/collator_side/validators_buffer.rs]: 135be9e81716a38fa55be442bb7839c06784057e

**Comment by slumber** [comment]
Note that there's also https://github.com/paritytech/polkadot/issues/5062 which is not addressed

Cleaning up a connection too soon may make things worse because we work in a low latency environment and collator's time window for producing and distributing a candidate is tight

On the other hand, if e.g. the availability is slow and the core stays occupied, we will stay connected to a group needlessly, but in this case validators will simply time us out which I don't think much of a deal.

File: node/network/collator-protocol/src/collator_side/mod.rs [hunk 5]
```diff
@@ -1023,6 +950,11 @@ pub(crate) async fn run<Context>(
 				FromOrchestra::Signal(Conclude) => return Ok(()),
 			},
 			(relay_parent, peer_id) = state.active_collation_fetches.select_next_some() => {
+				for authority_id in state.peer_ids.get(&peer_id).into_iter().flatten() {
+					// This peer is no longer interested in this relay parent.
+					state.validator_groups_buf.reset_validator_bit(relay_parent, authority_id);
```
Comment (eskimor) [inline node/network/collator-protocol/src/collator_side/mod.rs:955]: This is probably fine because we don't immediately take effect on this, but the above future also fires when the fetch was not yet concluded, but also when it exceeded its [unshared bandwith duration](https://github.com/paritytech/polkadot/blob/70bffc7365777089f6f84af524cb0c2938cbeddb/node/network/collator-protocol/src/collator_side/mod.rs#L643). Which means the fetch might be still in progress, would be bad to disconnect.

File: node/network/collator-protocol/src/collator_side/mod.rs [hunk 6]
```diff
@@ -541,13 +465,15 @@ async fn declare<Context>(ctx: &mut Context, state: &mut State, peer: PeerId) {
...
-/// revoke the previous connection request.
+/// Updates a set of connected validators based on their advertisement-bits
+/// in a validators buffer.
 #[overseer::contextbounds(CollatorProtocol, prefix = self::overseer)]
 async fn connect_to_validators<Context>(
 	ctx: &mut Context,
-	validator_ids: Vec<AuthorityDiscoveryId>,
+	validator_groups_buf: &ValidatorGroupsBuffer,
 ) {
+	let validator_ids = validator_groups_buf.validators_to_connect();
```
Comment (eskimor) [inline node/network/collator-protocol/src/collator_side/mod.rs:475]: `connect_to_validators` is actually only ever called on `distribute_collation`. This means while we perfectly keep track to whom we should stay connected, we don't make any use of it (as we will only update on the next collation we receive, which is way too late).

Comment (eskimor) [reply node/network/collator-protocol/src/collator_side/mod.rs:475]: There is actually already a separate issue about this: https://github.com/paritytech/polkadot/issues/5062 ... but it kind of has to be fixed with this PR, otherwise it won't improve anything.

File: node/network/collator-protocol/src/collator_side/validators_buffer.rs [hunk 7]
```diff
@@ -0,0 +1,287 @@
...
+			.enumerate()
+			.filter_map(|(idx, authority_id)| bits[idx].then_some(authority_id.clone()))
+			.collect()
+	}
+
+	/// Note a new collation distributed, setting this group validators bits to 1.
+	///
+	/// If max capacity is reached and the group is new, drops validators from the back
+	/// of the buffer.
+	pub fn note_collation_distributed(
```
[the remark this answers is not in the document]
Comment (slumber) [reply node/network/collator-protocol/src/collator_side/validators_buffer.rs:101]: We call it when distribute a collation to group, not when advertising to peer.

File: node/network/collator-protocol/src/collator_side/validators_buffer.rs [hunk 8]
```diff
@@ -0,0 +1,287 @@
...
+}
+
+/// Ring buffer of validator groups.
+///
+/// Tracks which peers we want to be connected to with respect to advertised collations.
+#[derive(Debug)]
+pub struct ValidatorGroupsBuffer {
+	buf: VecDeque<ValidatorsGroupInfo>,
+	validators: VecDeque<AuthorityDiscoveryId>,
+	should_be_connected: HashMap<Hash, BitVec>,
```
[the remark this answers is not in the document]
Comment (slumber) [reply node/network/collator-protocol/src/collator_side/validators_buffer.rs:65]: > Also, it is actually not clear at all why we need that mapping. Why are we keeping track of hashes here?

Collations are 1 to 1 mapped with relay parent without async backing, should be clarified in comments

File: node/network/collator-protocol/src/collator_side/mod.rs (hunk 6 above)
Comment (slumber) [reply node/network/collator-protocol/src/collator_side/mod.rs:475]: See https://github.com/paritytech/polkadot/pull/6022#issuecomment-1250214286

What this PR actually fixes is keeping connections to multiple groups alive on rotation boundaries, not clearing it up with previous or next group like it was done before.

We could ensure that we call `connect_to_validators` every (e.g.) 15 seconds or something, but I don't see much point in doing so.

File: node/network/collator-protocol/src/collator_side/mod.rs (hunk 5 above)
Comment (slumber) [reply node/network/collator-protocol/src/collator_side/mod.rs:955]: This timeout is supposed to be the upper limit isn't it? If the peer is slow there's no point in keeping the connection alive.

The case is a corner one anyway, what should happen is that we distribute a new collation right after the timeout is hit.

File: node/network/collator-protocol/src/collator_side/mod.rs (hunk 6 above)
Comment (slumber) [reply node/network/collator-protocol/src/collator_side/mod.rs:475]: By the way, the [test](https://github.com/paritytech/polkadot/blob/eab4fefd183125ca82e475635d721933636b9845/node/network/collator-protocol/src/collator_side/tests.rs#L994-L1093) very well expresses which behavior this PR is supposed to fix.

Comment (eskimor) [reply node/network/collator-protocol/src/collator_side/mod.rs:475]: Fair, but still - without that fix we actually don't ever disconnect at all (except on the next collation). We could have achieved that with less code. :grimacing: Anyhow, you are right - we can leave that for another PR.

File: node/network/collator-protocol/src/collator_side/mod.rs (hunk 5 above)
Comment (eskimor) [reply node/network/collator-protocol/src/collator_side/mod.rs]: The timeout hits after 150ms ... A peer might actually be quite fast and still hit the timeout. "Timeout" is in fact a bad naming here, because it isn't really a timeout - it is just, ok let's give one guy a head start so it can have unshared bandwidth for a while, so it can start validating asap. At the same time we don't want to delay propagating to other peers too long, so we start uploads one after the other in 150ms intervals. 

The general idea being: It is better if 2 validators get a full collation, than 5 validators getting 2/5 of a collation. Especially because if one validator has it all, it will start propagating to others as well, which is good for two reasons:

1. Multiplied bandwidth: This validator's bandwidth is added to the collators bandwidth for further propagation.
2. Potentially reduced latency: Let's assume that the validator that received the collation is a bit closer to one of the other validators than our collator. In that case it might have a smaller round trip time and can forward the collation faster, even if bandwidth is not actually the issue, but rather latency.

TL;DR: The timeout hitting, does neither mean that the validator is done, nor that it necessarily is bad. It is rather a normal condition that is planned to happen on everyday network conditions.

File: node/network/collator-protocol/src/collator_side/validators_buffer.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (eskimor) [reply node/network/collator-protocol/src/collator_side/validators_buffer.rs:108]: But the calling function is only called `distribute`, because it initiates distribution - by advertising. By the time we call `note_collation_distributed` - we only initiated advertising. I find this especially confusing, because once `distributed` ... which implies validators have fetched it, I would expect bits to get cleared not set.

[the remark this answers is not in the document]
Comment (slumber) [reply node/network/collator-protocol/src/collator_side/validators_buffer.rs:108]: agreed

File: node/network/collator-protocol/src/collator_side/mod.rs (hunk 5 above)
Comment (slumber) [reply node/network/collator-protocol/src/collator_side/mod.rs]: I can back-port changes from async backing branch where we actually catch a timeout:
https://github.com/paritytech/polkadot/blob/0b0c70a3f7b8839a06080d141c680e4104403eb1/node/network/collator-protocol/src/collator_side/mod.rs#L1212-L1221

And only disconnect if the collation was actually fetched

File: node/network/collator-protocol/src/collator_side/mod.rs (hunk 6 above)
Comment (slumber) [reply node/network/collator-protocol/src/collator_side/mod.rs:500]: e7fac145cc13e29e8b6e6a835b581874ae62d5b6

**Review by eskimor — APPROVED** [review]
Nice catch! Commit message could be better :see_no_evil:  ... this was not just a refactor, but a bug fix.

File: node/network/collator-protocol/src/lib.rs [hunk 9]
```diff
@@ -134,3 +137,23 @@ async fn modify_reputation(
...
+	if next_poll > now {
+		futures_timer::Delay::new(next_poll - now).await
+	}
+
+	Instant::now()
+}
+
+/// Returns an infinite stream that yields with an interval of `period`.
+fn tick_stream(period: Duration) -> impl FusedStream<Item = ()> {
+	futures::stream::unfold(Instant::now() + period, move |next_check| async move {
```
[the remark this answers is not in the document]
Comment (eskimor) [reply node/network/collator-protocol/src/lib.rs:155]: Not introduced here, but still wrong - isn't it?

[the remark this answers is not in the document]
Comment (slumber) [reply node/network/collator-protocol/src/lib.rs:155]: Yes, no need to add a period to the first timestamp

File: node/network/collator-protocol/src/collator_side/mod.rs [hunk 10]
```diff
@@ -1042,7 +1012,17 @@ pub(crate) async fn run<Context>(
 
 					send_collation(&mut state, next, receipt, pov).await;
 				}
-			}
+			},
+			_ = reconnect_stream.next() => {
+				let now = Instant::now();
+				if state
+					.last_connected_at
+					.map_or(false, |timestamp| now - timestamp > RECONNECT_TIMEOUT)
```
Comment (eskimor) [inline node/network/collator-protocol/src/collator_side/mod.rs:1020]: Why don't we just wait for `RECONNECT_TIMEOUT`? Instead of checking whether it elapsed in constant intervals?

Comment (slumber) [reply node/network/collator-protocol/src/collator_side/mod.rs:1020]: We ensure that we reconnect at least once every `RECONNECT_TIMEOUT` seconds

For example, if we just advertised a collation, there's no need to issue another reconnect request.

Comment (eskimor) [reply node/network/collator-protocol/src/collator_side/mod.rs:1020]: So we override the `Delay` future each time we receive an advertisment?

Comment (slumber) [reply node/network/collator-protocol/src/collator_side/mod.rs:1020]: We overwrite the timestamp on every call to `connect_to_validators`, yes
https://github.com/paritytech/polkadot/blob/ab57c1e02eb33b4b37ae8269c7d5f36eab4daa6b/node/network/collator-protocol/src/collator_side/mod.rs#L378-L379

Comment (eskimor) [reply node/network/collator-protocol/src/collator_side/mod.rs:1020]: Ok phrased differently: Why that inifinte stream of a delay shorter than what we actually need to wait for. Instead we could just have a `Delay` future that wakes exactly when needed by getting overridden at that very line?

Comment (slumber) [reply node/network/collator-protocol/src/collator_side/mod.rs:1020]: Probably not much difference, instead we wold juggle around pinned futures which doesn't look like a clean approach to me. We do some timeout checks on both sides of protocol so I decided to keep it consistent

Comment (eskimor) [reply node/network/collator-protocol/src/collator_side/mod.rs:1020]: Not sure I follow, but it also isn't that important right now.

**Comment by eskimor** [comment]
We need one more review on this one, @tdimitrov ?

**Review by tdimitrov — APPROVED** [review]

**Review by ordian — COMMENTED** [review]
> We do some timeout checks on both sides of protocol so I decided to keep it consistent

Just because things are implemented this way, doesn't mean they should. I'd like to see https://github.com/paritytech/polkadot/issues/4182 fixed sooner rather than later, rather than adding more things to refactor in the future.

The code changes look good, but I'd like to see this burned-in on Versi. You can create a companion PR in cumulus and use https://hub.docker.com/r/paritypr/polkadot-parachain-debug/tags.

**Comment by slumber** [comment]
@ordian unfortunately refactoring collator protocol right now is not the best idea, we will have a lot of changes with async backing.
This code was initially intended for async backing only, and the hack is present because it's adapted for master branch. In the feature branch we will simply reconnect on every activation.

**Comment by eskimor** [comment]
Only thing missing is Versi burnin. Or did we have one already?

**Comment by slumber** [comment]
> Only thing missing is Versi burnin. Or did we have one already?

on it

**Comment by slumber** [comment]
After a burn-in I discovered that 12-secs reconnect hack can cause an issue if collator doesn't see activations for some reason.
Since advertisements interest is not updated, collator will continuously attempt to reconnect to previously advertised validators.

Another option discussed with @eskimor was to reconnect on every activation. This solves the above issue in a way that
validator would simply time us out.
Also there's a 6 seconds timeout for fetching a collation, basically making it the same logic that would be used for async backing. This is needed to keep bits consistent.

**Comment by slumber** [comment]
> In case of a finality stall, relay chain block times would be (intentionally) high up to several minutes. Is that what you've observed?

Probably yes.
> > Another option discussed with @eskimor was to reconnect on every activation. This solves the above issue in a way that
> > validator would simply time us out.
> 
> Is that the same thing suggested in #4435 which we implemented and ended-up reverting?
> 
> If we're trying to solve #5062 here, that's not the way.

No, this PR introduces a state for tracking validators we want to be connected to, the question is **when** we want to connect.

Before the latest fix: Connect on advertisements + at least every 12 seconds

After the fix: Connect on advertisements + at each activation + timeout validators that don't fetch within 6 seconds

**Comment by eskimor** [comment]
I am not sure I understand the issue. What does continuously attempting to reconnect mean. Like at what level? Are we renewing connection requests constantly - if so where and how. Could you break down a bit more what you think is happening and why it is undesired?

**Comment by slumber** [comment]
See 2ed7f7a54924e9485cd0b84fe63002e6927e6e8d for details

**Comment by ordian** [comment]
Just to clarity: `connect_to_validators` doesn't issue a connection request per se. It adds validators to the reserved peerset. That means as long as we have them there and we're  connected to them, we'll try to connect (much more frequent than every 12 seconds). When we change the reserved set, e.g. set it to empty, we'll disconnect from the difference of old - new.

**Comment by slumber** [comment]
> Just to clarity: `connect_to_validators` doesn't issue a connection request per se. It adds validators to the reserved peerset. That means as long as we have them there and we're not connected to them, we'll try to connect (much more frequent than every 12 seconds). When we change the reserved set, e.g. set it to empty, we'll disconnect from the difference of old - new. If a validator is in our reversed, but he disconnects us, we'll still try to connect after disconnection.

Well then this fix doesn't really solve finality stall issue, thank you for clarification.
We could timeout validators if they don't fetch after 6 seconds and make sure we reconnect at least once every 12 seconds.

OR

Disconnect from everyone at least every 12 seconds.

**Comment by slumber** [comment]
> Uuh got it. Refreshing connections all the time on finality stall does not make too much sense indeed. So let's use leaf activation as refresh reason then.

this would still keep validators in reserved peerset as @ordian noted.

**Comment by eskimor** [comment]
Ok, when revisiting, I remember what I did not understand. Why is 3 happening? If we timeout after 12s the validator interest, we should disconnect - no?

**Comment by slumber** [comment]
> Ok, when revisiting, I remember what I did not understand. Why is 3 happening? If we timeout after 12s the validator interest, we should disconnect - no?

Previous approach is to connect at least once every 12 seconds based on the buffer without resetting the bits.

**Comment by slumber** [comment]
As a simple safeguard for master branch I [added](https://github.com/paritytech/polkadot/pull/6022/commits/6093b43ce934964f60d875dcfaca3795c768c254) a logic to remove all advertisements from the buffer on timeout.

This is correct because usually the advertisement goes out of view, if it doesn't and the timeout is hit (we had no new collations) there's no point in being connected to anyone, validators had plenty of time to fetch.

This **won't** work for async backing, collator can keep producing collations one after another and this timeout is never reached. Needs something smarter.

File: node/network/collator-protocol/src/collator_side/mod.rs [hunk 11]
```diff
@@ -1019,6 +1019,10 @@ pub(crate) async fn run<Context>(
 					.last_connected_at
 					.map_or(false, |timestamp| now - timestamp > RECONNECT_TIMEOUT)
 				{
+					// Remove all advertisements from the buffer if the timeout was hit.
+					// Usually, it shouldn't be necessary as leaves get deactivated, rather
+					// serves as a safeguard against finality lags.
+					state.validator_groups_buf.clear_advertisements();
```
Comment (eskimor) [inline node/network/collator-protocol/src/collator_side/mod.rs:1025]: Wait a second. This looks wrong. Doesn't this clear all advertisments? Shouldn't this be restricted to advertisments that are older than 12s?

Comment (slumber) [reply node/network/collator-protocol/src/collator_side/mod.rs:1025]: If we had an advertisement less than 12 seconds old, we would've called `distribute_collation` and update `last_connected_at`. Thus, this is only hit if we had no new collations in last 12 seconds
